rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+
+![Flamegrill - Results Summary](theme-shape/perf-flamegrill.png)
+
+## Proposed changes to the theme shape
+
+1. [Inline values in alias tokens](#inline-alias)
+2. [Do not inject global color tokens as CSS variables](#no-global-tokens)
+3. [Merge global and alias tokens to a single object](#merge-global-alias)
+4. [Flatten the tokens object](#flatten-tokens)
+5. [Hash variable names](#hash-variables)
+6. [Be cautious when adding new tokens to the theme](#be-cautious)
+7. [Get rid of all CSS variables](#no-css-vars)
+
+### <a name="inline-alias"></a>Inline values in alias tokens
+
+Global tokens reference values, alias tokens reference the global tokens:
+
+```css
+/* Before */
+--global-red: #f00;
+--alias-error: var(--global-red);
+```
+
+Performance analysis shows, that by inlining the values directly into CSS variables for alias tokens we can improve Rendering perf.
+
+Proposed change:
+
+```css
+/* After */
+--global-red: #f00; /* no change here */
+--alias-error: #f00; /* inlined value from --global-red */
+```
+
+This change has already been implemented in [#19660](https://github.com/microsoft/fluentui/pull/19660).
+
+### <a name="no-global-tokens"></a>Do not inject global color tokens as CSS variables
+
+The main reason why CSS variables are used is to be able to switch themes just by changing the CSS variables values.
+
+But, by design, global tokens never change their values when switching theme, so there is no point to have them as CSS variables.
+
+The proposal is to **remove global color tokens from the theme object** and **do not inject them to DOM as CSS variables**.
+
+As shown in the [Performance analysis](#performance-analysis) analysis, this reduces the number of CSS variables by a half and significantly improves rendering performance.
+
+#### What if I need global color tokens in component styles or overrides
+
+First of all, using a global color token in styles directly is almost always incorrect as the color does not change when switching themes.
+
+For the rare cases where the global color token is needed, `react-theme` will export all the global tokens.
+
+```js
+/* Before */
+import { makeStyles } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  root: theme => ({
+    color: theme.alias.color.neutral.neutralStroke1, // color: var(--alias-color-neutral-neutralStroke1)
+    background: theme.global.palette.grey[50], // background: var(--global-palette-grey-50)
+  }),
+});
+```
+
+```js
+/* After */
+import { makeStyles, globalColorTokens } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  root: theme => ({
+    color: theme.alias.color.neutral.neutralStroke1, // color: var(--alias-color-neutral-neutralStroke1)
+    background: globalColorTokens.palette.grey[50], // background: '#ddd' 👈 value inlined/bound during babel-plugin/build
```

Comment:
If importing globalColorTokens is possible because the value never changes (which is correct), then look also at the alias tokens on `theme`. Their value is also static between all them changes since they simply convert a dot annotation to a css variable string. The `var()` never changes between theme switches.

Thus, we can actually get rid of the `theme => ({})` functions altogether and have only a single imported theme object. This would allow tree shaking theme aliases in addition to globals. We would also reduce the API surface as makeStyles would only need to support objects. We would then also save all the function calls.

The caveat is that it forces us to rely on css variables as values in makeStyles. We could not fallback to static values (`#000` vs `var(--foreground1)`) because the value would need to update on theme switch but it could not because it is imported.